### PR TITLE
Allow fetching translation keys from validation messages

### LINKED ISSUE #4455: Translation keys can't be specified for default validation functions
### Elixir version

1.17.0

### Database and Version

PostgreSQL 16.3

### Ecto Versions

3.11.2

### Database Adapter and Versions (postgrex, myxql, etc)

postgrex 0.18.0

### Current behavior

Specifying a message with translation keys does not work with the built-in validation functions (anything that calls `validate_change` and accepts a `message` option). E.g. attempting to use this:
```ex
  changeset
  |> put_change(:mime, mime)
  |> validate_inclusion(:mime, ~W(image/jpeg image/png),  message: {"(%{value}) is invalid", value: mime})
```
raises an error:
```
** (FunctionClauseError) no function clause matching in anonymous fn/1 in Ecto.Changeset.validate_change/3
    (ecto 3.11.2) lib/ecto/changeset.ex:2430: anonymous fn({:mime, {{"%{value} is not allowed", [value: "application/pdf"]}, [validation: :inclusion, enum: ["image/jpeg", "image/png"]]}}) in Ecto.Changeset.validate_change/3
    (elixir 1.17.0) lib/enum.ex:1703: Enum."-map/2-lists^map/1-1-"/2
    (ecto 3.11.2) lib/ecto/changeset.ex:2430: Ecto.Changeset.validate_change/3
```

### Expected behavior

Built-in validation functions should allow messages to include translation keys, similarly to `add_error/4`.

### PR DESCRIPTION
**Pull request by liamwhite**
Fixes #4455, with the caveat that there is now a non-uniform API for messages between validations (which can accept these) and constraints (which cannot accept these).

### CODE REVIEW

**Review by josevalim — APPROVED** [review]

File: lib/ecto/changeset.ex [hunk 1]
```diff
@@ -3120,6 +3185,8 @@ defmodule Ecto.Changeset do
       * "must be greater than or equal to %{number}"
       * "must be equal to %{number}"
       * "must be not equal to %{number}"
+      Can also be a `{msg, opts}` tuple, to provide additional options
+      when using`traverse_errors/2`.
```
Comment (greg-rychlewski) [inline lib/ecto/changeset.ex:3189]:
```suggestion
      when using `traverse_errors/2`.
```

File: lib/ecto/changeset.ex [hunk 2]
```diff
@@ -3235,7 +3304,9 @@ defmodule Ecto.Changeset do
 
   ## Options
 
-    * `:message` - the message on failure, defaults to "does not match confirmation"
+    * `:message` - the message on failure, defaults to "does not match confirmation".
+      Can also be a `{msg, opts}` tuple, to provide additional options
+      when using`traverse_errors/2`.
```
Comment (greg-rychlewski) [inline lib/ecto/changeset.ex:3309]:
```suggestion
      when using `traverse_errors/2`.
```

File: lib/ecto/changeset.ex [hunk 3]
```diff
@@ -3305,7 +3385,9 @@ defmodule Ecto.Changeset do
 
   ## Options
 
-    * `:message` - the message on failure, defaults to "must be accepted"
+    * `:message` - the message on failure, defaults to "must be accepted".
+      Can also be a `{msg, opts}` tuple, to provide additional options
+      when using`traverse_errors/2`.
```
Comment (greg-rychlewski) [inline lib/ecto/changeset.ex:3390]:
```suggestion
      when using `traverse_errors/2`.
```

File: lib/ecto/changeset.ex [hunk 4]
```diff
@@ -3288,11 +3358,21 @@ defmodule Ecto.Changeset do
...
-      do: [{error_field, {message(opts, "can't be blank"), [validation: :required]}}],
+      do: [{error_field, message(opts, "can't be blank", validation: :required)}],
       else: []
   end
 
-  defp message(opts, key \\ :message, default) do
+  defp message(opts, key \\ :message, default, message_opts) do
+    case Keyword.get(opts, key, default) do
+      {message, extra_opts} when is_binary(message) and is_list(extra_opts) ->
+        {message, Keyword.merge(message_opts, extra_opts)}
```
Comment (greg-rychlewski) [inline lib/ecto/changeset.ex:3368]: Should we let users override the ones being set by Ecto? I'm not sure if they are public and other libraries are depending on them.

Comment (josevalim) [reply lib/ecto/changeset.ex:3368]: They are always public and I think overriding is fine. :)

**Comment by josevalim** [comment]
Merging as @greg-rychlewski's feedback has been addressed now!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**